src/chainstate/stacks/index/storage.rs

```diff
@@ -0,0 +1,1213 @@
+       
+        // update internal structures
+        self.cur_block = bhh.clone();
+        self.cur_block_fd = None;
+        self.cur_block_fork_ptr = None;
+
+        self.last_extended = Some(bhh.clone());
+        self.last_extended_trie = Some(trie_buf);
+                
+        trace!("Extended to {:?} in {:?}", &self.cur_block, &block_path);
+        Ok(())
+    }
+
+    pub fn open_block(&mut self, bhh: &BlockHeaderHash, readwrite: bool) -> Result<(), Error> {
+        let block_fork_ptr = self.fork_table.get_fork_ptr(bhh)?;
+
+        if Some(*bhh) == self.last_extended {
+            // nothing to do -- we're already ready.
+            // just clear out.
+            self.cur_block_fd = None;
+            self.cur_block = bhh.clone();
+            self.cur_block_fork_ptr = Some(block_fork_ptr);
+            self.readonly = !readwrite;
+            return Ok(());
+        }
+
+        // opening a different Trie than the one we're extending
+        let block_path = self.cached_block_path(bhh);
+        let fd = fs::OpenOptions::new()
+                    .read(true)
+                    .write(readwrite)
+                    .open(&block_path)
+                    .map_err(|e| {
+                        if e.kind() == io::ErrorKind::NotFound {
+                            error!("File not found: {:?}", &block_path);
+                            Error::NotFoundError
+                        }
+                        else {
+                            Error::IOError(e)
+                        }
+                    })?;
+
+        self.cur_block = bhh.clone();
+        self.cur_block_fd = Some(fd);
+        self.cur_block_fork_ptr = Some(block_fork_ptr);
+        self.readonly = !readwrite;
+        Ok(())
+    }
+    
+    pub fn get_cur_block(&self) -> BlockHeaderHash {
+        self.cur_block.clone()
+    }
+    
+    pub fn block_walk(&mut self, back_block: u32) -> Result<BlockHeaderHash, Error> {
+        trace!("block_walk from {:?} back {}", &self.cur_block, back_block);
+
+        let prev_block = match self.cur_block_fork_ptr {
+            Some(ref fork_ptr) => {
+                self.fork_table.walk_back_from(fork_ptr, &self.cur_block, back_block)?
+            },
+            None => {
+                let fork_ptr = self.fork_table.get_fork_ptr(&self.cur_block)?;
+                self.cur_block_fork_ptr = Some(fork_ptr.clone());
+                self.fork_table.walk_back_from(&fork_ptr, &self.cur_block, back_block)?
+            }
+        };
+        
+        if prev_block == TrieFileStorage::block_sentinel() {
+            trace!("Not found: {:?} back {}", &self.cur_block, back_block);
+            return Err(Error::NotFoundError);
+        }
+        
+        trace!("block_walk from {:?} back {} is {:?}", &self.cur_block, back_block, &prev_block);
+        Ok(prev_block)
+    }
+    
+    pub fn root_ptr(&self) -> u32 {
+        if Some(self.cur_block) == self.last_extended {
```

Comment:
I think I understand what's happening here, but I'm not positive, so I want to test my understanding --

in the case that `self.cur_block` is the same as `self.last_extended`, that means that the current block is actually loaded in memory, in the form of a TrieRAM. In memory, the contents of the trie start at byte _0_ -- it doesn't keep the parent's hash in memory.

however, if they aren't equal, it means it'll be servicing reads from disk, where the parent hash is stored at byte 0, so you have to begin reading at byte 32?

Reply:
That's correct.  The trie file stores the parent hash in bytes 0-31, and the nodes at 32+ (the root node itself is at byte 32).